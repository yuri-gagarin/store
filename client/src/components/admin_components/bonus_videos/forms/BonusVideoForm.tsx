import React, { useState, useEffect, useRef }  from "react";
import { Button, Form, TextArea } from "semantic-ui-react";

export type FormState = {
  description: string;
  youTubeURL: string;
  vimeoURL: string;
}

interface Props {
  description: string;
  youTubeURL: string;
  vimeoURL: string;
  handleCreateProduct(data: FormState): void;
  handleUpdateProduct(data: FormState): void;
}


const ProductForm: React.FC<Props> = ({ description, youTubeURL, vimeoURL, handleCreateProduct, handleUpdateProduct }): JSX.Element => {
  
  const [ newForm, setNewForm ] = useState<boolean>(true)
  const [ formState, setFormState ] = useState<FormState>({ description, youTubeURL, vimeoURL });
  const bonusVideoFormRef = useRef<HTMLDivElement>(document.createElement("div"));
  
  const handleYouTubeURLChange = (e: React.ChangeEvent<HTMLInputElement>):void => {
    setFormState({
      ...formState,
      youTubeURL: e.target.value
    });
  };
  const handleVimeoURLChange = (e: React.ChangeEvent<HTMLInputElement>):void => {
    setFormState({
      ...formState,
      vimeoURL: e.target.value
    });
  };
  const hadnleDescriptionChange = (e: React.ChangeEvent<HTMLInputElement>): void => {
    setFormState({
      ...formState,
      description: e.target.value
    });
  };
  
  const handleSubmit = (e: React.MouseEvent<HTMLButtonElement>) => {
    e.preventDefault();
    if (newForm) {
      handleCreateProduct(formState);
    } else {
      handleUpdateProduct(formState);
    }
  };

  useEffect(() => {
    if (description || youTubeURL || vimeoURL) {
      setNewForm(false);
    } else {
      setNewForm(true);
    }
  },  [description, youTubeURL, vimeoURL]);

  useEffect(() => {
    if (bonusVideoFormRef.current) {
      const elem = bonusVideoFormRef.current.getBoundingClientRect();
      window.scrollTo({
        top: elem.bottom,
        behavior: "smooth"
      })
    }
  }, [bonusVideoFormRef]);
  
  return (
    <div className="createProductFormHolder" ref={bonusVideoFormRef}>
      <Form id="createProductForm">
        <Form.Field>
          <label>Video YouTube URL:</label>
          <input 
            onChange={handleYouTubeURLChange} 
            placeholder="YouTube URL here ..." 
            value={formState.youTubeURL}
          />
        </Form.Field>
        <Form.Field>
          <label>Video Vimeo URL:</label>
          <input 
            onChange={handleVimeoURLChange} 
            placeholder="Vimeo URL here..." 
            value={formState.vimeoURL}
          />
        </Form.Field>
        <Form.Field
          id='form-textarea-control-opinion'
          control={TextArea}
          label='Video Description'
          onChange={hadnleDescriptionChange}
          placeholder='Video description here...'
          value={formState.description}
         />
         {
           newForm 
            ? <Button type='submit' onClick={handleSubmit} content= "Save New Bonus Video" />
            : <Button type='submit' onClick={handleSubmit} content= "Update Bonus Video" />
         }  
      </Form>
    </div>
  );
};

export default ProductForm;