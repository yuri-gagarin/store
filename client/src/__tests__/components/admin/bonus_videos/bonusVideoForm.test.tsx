import React from "react";
import BonusVideoForm from "../../../../components/admin_components/bonus_videos/forms/BonusVideoForm";
import { mount, ReactWrapper } from "enzyme";
// helpers //

describe("BonusVideo Form Component render tests", () => {
  describe("Empty Form render tests", () => {
    let wrapper: ReactWrapper;

    beforeAll(() => {
      window.scrollTo = jest.fn;
      // mock data //
      const bonusVideoData: IBonusVideoData = {
        _id: "",
        youTubeURL: "",
        vimeoURL: "",
        description: "",
        createdAt: ""
      }
      // mount form //
      wrapper = mount(
        <BonusVideoForm  
          bonusVideoData={bonusVideoData} 
          handleCreateBonusVideo={jest.fn}
          handleUpdateBonusVideo={jest.fn}
        />
      );
    });

    it("Should properly mount", () => {
      expect(wrapper).toMatchSnapshot();
    });
    it("Should render '#bonusVideoYoutubeInput'", () => {
      expect(wrapper.render().find("#bonusVideoYoutubeInput").length).toEqual(1);
    });
    it("Should render '#bonusVideoVimeoInput'", () => {
      expect(wrapper.render().find("#bonusVideoVimeoInput").length).toEqual(1);
    });
    it("Should render '#bonusVideoDescInput'", () => {
      expect(wrapper.render().find("#bonusVideoDescInput").length).toEqual(1);
    });
    it("Should render '#bonusVideoFormCreate' Button", () => {
      expect(wrapper.render().find("#bonusVideoFormCreate").length).toEqual(1);
    });
    it("Should NOT render '#bonusVideoFormUpdate' Button'", () => {
      expect(wrapper.render().find("#bonusVideoFormUpdate").length).toEqual(0);
    });
  });
  
  describe("Form with valid data render tests", () => {
    let wrapper: ReactWrapper; let mockData: IBonusVideoData;

    beforeAll(() => {
      window.scrollTo = jest.fn;
      // mock video data //
      mockData = {
        _id: "1111",
        youTubeURL: "url",
        vimeoURL: "url",
        description: "desc",
        createdAt: "now"
      };

      wrapper = mount(
        <BonusVideoForm 
          bonusVideoData={mockData}
          handleCreateBonusVideo={jest.fn}
          handleUpdateBonusVideo={jest.fn}
        />
      );
    });

    it("Should properly mount", () => {
      expect(wrapper).toMatchSnapshot();
    });
    it("Should render '#bonusVideoYoutubeInput'", () => {
      expect(wrapper.render().find("#bonusVideoYoutubeInput").length).toEqual(1);
    });
    it("Should set correct data in '#bonusVideoYoutubeInput'", () => {
      const input = wrapper.render().find("#bonusVideoYoutubeInput");
      expect(input.val()).toEqual(mockData.youTubeURL);
    });
    it("Should render '#bonusVideoVimeoInput'", () => {
      expect(wrapper.render().find("#bonusVideoVimeoInput").length).toEqual(1);
    });
    it("Should set correct data in '#bonusVideoVimeoInput'", () => {
      const input = wrapper.render().find("#bonusVideoVimeoInput");
      expect(input.val()).toEqual(mockData.vimeoURL);
    });
    it("Should render '#bonusVideoDescInput'", () => {
      expect(wrapper.render().find("#bonusVideoDescInput").length).toEqual(1);
    });
    it("Should set correct data in '#bonusVideoDescInput'", () => {
      const input = wrapper.render().find("#bonusVideoDescInput");
      expect(input.val()).toEqual(mockData.description);
    });
    it("Should NOT render '#bonusVideoFormCreate' Button", () => {
      expect(wrapper.render().find("#bonusVideoFormCreate").length).toEqual(0);
    });
    it("Should NOT render '#bonusVideoFormUpdate' Button'", () => {
      expect(wrapper.render().find("#bonusVideoFormUpdate").length).toEqual(1);
    });
  });
})