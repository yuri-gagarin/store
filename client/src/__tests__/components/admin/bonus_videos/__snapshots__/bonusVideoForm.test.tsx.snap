// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BonusVideo Form Component render tests Empty Form render tests Should properly mount 1`] = `
<BonusVideoForm
  description=""
  handleCreateBonusVideo={[Function]}
  handleUpdateBonusVideo={[Function]}
  newForm={true}
  vimeoURL=""
  youTubeURL=""
>
  <div
    id="createBonusVideoForm"
  >
    <Form
      as="form"
    >
      <form
        className="ui form"
        onSubmit={[Function]}
      >
        <FormField>
          <div
            className="field"
          >
            <label>
              Video YouTube URL:
            </label>
            <input
              id="bonusVideoYoutubeInput"
              onChange={[Function]}
              placeholder="YouTube URL here ..."
              value=""
            />
          </div>
        </FormField>
        <FormField>
          <div
            className="field"
          >
            <label>
              Video Vimeo URL:
            </label>
            <input
              id="bonusVideoVimeoInput"
              onChange={[Function]}
              placeholder="Vimeo URL here..."
              value=""
            />
          </div>
        </FormField>
        <FormField
          control={[Function]}
          id="bonusVideoDescInput"
          label="Video Description"
          onChange={[Function]}
          placeholder="Video description here..."
          value=""
        >
          <div
            className="field"
          >
            <label
              htmlFor="bonusVideoDescInput"
            >
              Video Description
            </label>
            <TextArea
              aria-describedby={null}
              as="textarea"
              id="bonusVideoDescInput"
              onChange={[Function]}
              placeholder="Video description here..."
              rows={3}
              value=""
            >
              <Ref
                innerRef={
                  Object {
                    "current": <textarea
                      id="bonusVideoDescInput"
                      placeholder="Video description here..."
                      rows="3"
                    />,
                  }
                }
              >
                <RefFindNode
                  innerRef={
                    Object {
                      "current": <textarea
                        id="bonusVideoDescInput"
                        placeholder="Video description here..."
                        rows="3"
                      />,
                    }
                  }
                >
                  <textarea
                    aria-describedby={null}
                    id="bonusVideoDescInput"
                    onChange={[Function]}
                    onInput={[Function]}
                    placeholder="Video description here..."
                    rows={3}
                    value=""
                  />
                </RefFindNode>
              </Ref>
            </TextArea>
          </div>
        </FormField>
        <Button
          as="button"
          content="Save New Bonus Video"
          id="bonusVideoFormCreate"
          onClick={[Function]}
          type="submit"
        >
          <Ref
            innerRef={
              Object {
                "current": <button
                  class="ui button"
                  id="bonusVideoFormCreate"
                  type="submit"
                >
                  Save New Bonus Video
                </button>,
              }
            }
          >
            <RefFindNode
              innerRef={
                Object {
                  "current": <button
                    class="ui button"
                    id="bonusVideoFormCreate"
                    type="submit"
                  >
                    Save New Bonus Video
                  </button>,
                }
              }
            >
              <button
                className="ui button"
                id="bonusVideoFormCreate"
                onClick={[Function]}
                type="submit"
              >
                Save New Bonus Video
              </button>
            </RefFindNode>
          </Ref>
        </Button>
      </form>
    </Form>
  </div>
</BonusVideoForm>
`;

exports[`BonusVideo Form Component render tests Form with valid data render tests Should properly mount 1`] = `
<BonusVideoForm
  description="desc"
  handleCreateBonusVideo={[Function]}
  handleUpdateBonusVideo={[Function]}
  newForm={false}
  vimeoURL="url"
  youTubeURL="url"
>
  <div
    id="createBonusVideoForm"
  >
    <Form
      as="form"
    >
      <form
        className="ui form"
        onSubmit={[Function]}
      >
        <FormField>
          <div
            className="field"
          >
            <label>
              Video YouTube URL:
            </label>
            <input
              id="bonusVideoYoutubeInput"
              onChange={[Function]}
              placeholder="YouTube URL here ..."
              value="url"
            />
          </div>
        </FormField>
        <FormField>
          <div
            className="field"
          >
            <label>
              Video Vimeo URL:
            </label>
            <input
              id="bonusVideoVimeoInput"
              onChange={[Function]}
              placeholder="Vimeo URL here..."
              value="url"
            />
          </div>
        </FormField>
        <FormField
          control={[Function]}
          id="bonusVideoDescInput"
          label="Video Description"
          onChange={[Function]}
          placeholder="Video description here..."
          value="desc"
        >
          <div
            className="field"
          >
            <label
              htmlFor="bonusVideoDescInput"
            >
              Video Description
            </label>
            <TextArea
              aria-describedby={null}
              as="textarea"
              id="bonusVideoDescInput"
              onChange={[Function]}
              placeholder="Video description here..."
              rows={3}
              value="desc"
            >
              <Ref
                innerRef={
                  Object {
                    "current": <textarea
                      id="bonusVideoDescInput"
                      placeholder="Video description here..."
                      rows="3"
                    >
                      desc
                    </textarea>,
                  }
                }
              >
                <RefFindNode
                  innerRef={
                    Object {
                      "current": <textarea
                        id="bonusVideoDescInput"
                        placeholder="Video description here..."
                        rows="3"
                      >
                        desc
                      </textarea>,
                    }
                  }
                >
                  <textarea
                    aria-describedby={null}
                    id="bonusVideoDescInput"
                    onChange={[Function]}
                    onInput={[Function]}
                    placeholder="Video description here..."
                    rows={3}
                    value="desc"
                  />
                </RefFindNode>
              </Ref>
            </TextArea>
          </div>
        </FormField>
        <Button
          as="button"
          content="Update Bonus Video"
          id="bonusVideoFormUpdate"
          onClick={[Function]}
          type="submit"
        >
          <Ref
            innerRef={
              Object {
                "current": <button
                  class="ui button"
                  id="bonusVideoFormUpdate"
                  type="submit"
                >
                  Update Bonus Video
                </button>,
              }
            }
          >
            <RefFindNode
              innerRef={
                Object {
                  "current": <button
                    class="ui button"
                    id="bonusVideoFormUpdate"
                    type="submit"
                  >
                    Update Bonus Video
                  </button>,
                }
              }
            >
              <button
                className="ui button"
                id="bonusVideoFormUpdate"
                onClick={[Function]}
                type="submit"
              >
                Update Bonus Video
              </button>
            </RefFindNode>
          </Ref>
        </Button>
      </form>
    </Form>
  </div>
</BonusVideoForm>
`;
